src/unix/notbsd/linux/other/mod.rs

```diff
@@ -1,5 +1,110 @@
 pub type __priority_which_t = ::c_uint;
 
+cfg_if! {
+    if #[cfg(libc_union)] {
+        #[repr(C)]
+        #[derive(Copy, Clone)]
+        struct addr_bnd_t {
+            lower: *mut ::c_void,
+            upper: *mut ::c_void,
+        }
+
+        #[repr(C)]
+        #[derive(Copy, Clone)]
+        union sigfault_t_anonymous_union {
+            addr_bnd: addr_bnd_t,
+            pkey: u32,
+        }
+
+        #[repr(C)]
+        #[derive(Copy, Clone)]
+        struct sigfault_t {
+            addr: *mut ::c_void,
+            #[cfg(target_arch = "sparc")]
+            trapno: ::c_int,
+            addr_lsb: ::c_short,
+            anonymous_union: sigfault_t_anonymous_union,
+        }
+
+        #[repr(C, packed)]
+        #[derive(Copy, Clone)]
+        union sifields_t {
+            _pad: [::c_int; 29],
+            sigfault: sigfault_t,
+        }
+
+        impl ::fmt::Debug for sigfault_t_anonymous_union {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                // Excludes the fields because we can't know which is valid
+                f.debug_struct("sigfault_t_anonymous_union")
+                    .finish()
+            }
+        }
+
+        impl ::fmt::Debug for sigfault_t {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                // TODO: include trapno on Sparc
+                f.debug_struct("sigfault_t")
+                    .field("addr", &self.addr)
+                    .field("addr_lsb", &self.addr_lsb)
+                    .field("anonymous_union", &self.anonymous_union)
+                    .finish()
+            }
+        }
+
+        impl ::fmt::Debug for sifields_t {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                // No way to print anything more detailed without the
+                // discriminant from siginfo_t
+                f.debug_struct("sifields_t").finish()
+            }
+        }
+    } else {
+        type sifields_t = [::c_int; 29];
+    }
+}
+
+s_no_extra_traits! {
+    pub struct siginfo_t {
+        pub si_signo: ::c_int,
+        pub si_errno: ::c_int,
+        pub si_code: ::c_int,
+        sifields: sifields_t,
+
+        #[cfg(target_arch = "x86_64")]
+        _align: [u64; 0],
+        #[cfg(not(target_arch = "x86_64"))]
+        _align: [usize; 0],
+    }
+}
+
+cfg_if! {
+    if #[cfg(libc_union)] {
+        impl siginfo_t {
+            pub unsafe fn si_addr(&self) -> *mut ::c_void {
+                self.sifields.sigfault.addr
+            }
+        }
+    } else {
+        impl siginfo_t {
+            pub unsafe fn si_addr(&self) -> *mut ::c_void {
+                *(self.sifields.as_ptr() as *const *mut ::c_void)
```

Comment:
Yeah, this isn't correct. There should be 4 bytes of padding between `si_code` and `sifields`, but you won't get that since `[i32; 29]` has an alignment of 4.

Is there a sane way to fix this?